---
title: "Case study 1: Breast cancer"
output:
  rmdformats::readthedown:
    self_contained: true
    thumbnails: true
    lightbox: true
    gallery: false
    highlight: tango
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE, warning = FALSE, message = FALSE, tidy = TRUE)

WhereAmI <- "~/Dropbox/PROOF/Manuscript/mixOmics/diablo/analyses/casestudy1_brca/"

## load libraries
require(parallel)
library(knitr)
library(cowplot)
library(tidyverse)
library(mixOmics)
library(amritr)
library(corrplot)
require(parallel)
library(sear)
library(UpSetR)
library(venn)
library(igraph)
library(ggraph)
library(ggalt)
library(RColorBrewer)

## load data
load("~/Dropbox/PROOF/Manuscript/mixOmics/diablo_datasets/brca/trainTestDatasetsNormalized.RDATA")
source("~/Dropbox/PROOF/Manuscript/mixOmics/diablo/analyses/casestudy1_brca/brcaFunctions.R")
```

# Analysis 1: Comparison of classification performance between DIABLO models and other classification schemes (Concatecation and Ensemble) - only using mRNA, miRNA and CpGs

## Number of samples and variables per dataset

```{r, results = 'hide'}
Y.train <- droplevels(pam50Train0$Call)
names(Y.train) <- rownames(pam50Train0)
table(Y.train)
X.train <- list(mRNA = mrnaTrain0, miRNA = mirnaTrain0, CpGs = methTrain0, Proteins = protTrain0)
all(names(Y.train) == rownames(X.train[[1]]))
all(names(Y.train) == rownames(X.train[[2]]))
all(names(Y.train) == rownames(X.train[[3]]))
all(names(Y.train) == rownames(X.train[[4]]))

Y.test <- droplevels(pam50Test0$Call)
names(Y.test) <- rownames(pam50Test0)
table(Y.test)
X.test <- list(mRNA = mrnaTest0, miRNA = mirnaTest0, CpGs = methTest0)
all(names(Y.test) == rownames(X.test[[1]]))
all(names(Y.test) == rownames(X.test[[2]]))
all(names(Y.test) == rownames(X.test[[3]]))
```

```{r}
sapply(X.train, dim) %>% 
  as.data.frame %>% 
  mutate(Attribute = c("Number of Samples", "Number of features"))
```

## Phenotype breakdown

```{r}
table(Y.train)
```

## Tune DIABLO models

```{r}
# folds <- 5; nrepeat <- 5; cpus <- 8;
# ## null design
# designnull <- matrix(0, nrow = length(X.test), ncol = length(X.test))
# rownames(designnull) <- colnames(designnull) <- names(X.test)
# test.keepX = list(mRNA = c(2, 5, 10, 15, 20), miRNA = c(2, 5, 10, 15, 20), CpGs = c(2, 5, 10, 15, 20))
# ncomp = 3
# t1 <- proc.time()
# tune = tune.block.splsda(X = X.train[names(X.test)], Y = Y.train, ncomp = ncomp, test.keepX = test.keepX, dist = "centroids.dist",
#                           design = designnull, validation = "Mfold", folds = folds, nrepeat = nrepeat, cpus = cpus)
# t2 <- proc.time()
# (t2-t1)/60 # 39.5 minutes
# saveRDS(tune, "~/Dropbox/PROOF/Manuscript/mixOmics/diablo_datasets/brca/brcaPanel_tune_nullDesign.rds")
diablonullDesign <- readRDS("~/Dropbox/PROOF/Manuscript/mixOmics/diablo_datasets/brca/brcaPanel_tune_nullDesign.rds")

## full design
# designfull <- matrix(1, nrow = length(X.test), ncol = length(X.test))
# rownames(designfull) <- colnames(designfull) <- names(X.test)
# diag(designfull) <- 0
# t1 <- proc.time()
# tune = tune.block.splsda(X = X.train[names(X.test)], Y = Y.train, ncomp = ncomp, test.keepX = test.keepX, dist = "centroids.dist",
#                           design = designfull, validation = "Mfold", folds = folds, nrepeat = nrepeat, cpus = cpus)
# t2 <- proc.time()
# (t2-t1)/60 # 42.1 minutes
# saveRDS(tune, "~/Dropbox/PROOF/Manuscript/mixOmics/diablo_datasets/brca/brcaPanel_tune_fullDesign.rds")
diablofullDesign <- readRDS("~/Dropbox/PROOF/Manuscript/mixOmics/diablo_datasets/brca/brcaPanel_tune_fullDesign.rds")
```

## estimate DIABLO train and test error rates

```{r}
# ncomp = 3;
# X_train <- X.train[names(X.test)]
# y_train <- Y.train
# X_test <- X.test
# y_test <- Y.test
# folds <- 5; iter <- 5;
# 
# ## DIABLO null
# keepX = diablonullDesign$choice.keepX[names(X.test)]
# diablonull_perf <- diabloFit(keepX, ncomp, X_train, X_test, y_train, y_test, design=designnull, folds, iter, modName="DIABLO_null")
# saveRDS(diablonull_perf, "~/Dropbox/PROOF/Manuscript/mixOmics/diablo_datasets/brca/diablonull_perf.rds")
diablonull_perf <- readRDS("~/Dropbox/PROOF/Manuscript/mixOmics/diablo_datasets/brca/diablonull_perf.rds")

# ## DIABLO full
# keepX = diablofullDesign$choice.keepX[names(X.test)]
# diablofull_perf <- diabloFit(keepX, ncomp, X_train, X_test, y_train, y_test, design=designfull, folds, iter, modName="DIABLO_full")
# saveRDS(diablofull_perf, "~/Dropbox/PROOF/Manuscript/mixOmics/diablo_datasets/brca/diablofull_perf.rds")
diablofull_perf <- readRDS("~/Dropbox/PROOF/Manuscript/mixOmics/diablo_datasets/brca/diablofull_perf.rds")
```

## Tune sPLSDA models

```{r}
# Concatenation
# X_train <- do.call(cbind, X.train[names(X.test)])
# colnames(X_train) <- paste(rep(names(X.test), sapply(X.test, ncol)), unlist(sapply(X.test, colnames)), sep="_")
# y_train <- Y.train
# t1 <- proc.time()
# tune= tune(method = "splsda", X_train, y_train, ncomp=ncomp, logratio="none",
#     test.keepX = c(2, 5, 10, 15, 20), dist="centroids.dist", progressBar = TRUE, folds = folds, nrepeat = nrepeat)
# t2 <- proc.time()
# (t2-t1)/60 # 5.6 minutes
# saveRDS(tune, "~/Dropbox/PROOF/Manuscript/mixOmics/diablo_datasets/brca/splsdaConcat.rds")
splsdaConcat <- readRDS("~/Dropbox/PROOF/Manuscript/mixOmics/diablo_datasets/brca/splsdaConcat.rds")

# Ensemble
# X_train <- X.train[names(X.test)]
# y_train <- Y.train
# X_test <- X.test
# y_test <- Y.test
# t1 <- proc.time()
# tune_singleOmics_splsda <- lapply(X_train, function(i){
#   tune(method = "splsda", i, y_train, ncomp=ncomp, logratio="none",
#     test.keepX = c(2, 5, 10, 15, 20), dist="centroids.dist", progressBar = FALSE, folds = folds, nrepeat = nrepeat)
# })
# t2 <- proc.time()
# (t2-t1)/60 # 6.1 minutes
# saveRDS(tune_singleOmics_splsda, "~/Dropbox/PROOF/Manuscript/mixOmics/diablo_datasets/brca/tune_singleOmics_splsda.rds")
tune_singleOmics_splsda <- readRDS("~/Dropbox/PROOF/Manuscript/mixOmics/diablo_datasets/brca/tune_singleOmics_splsda.rds")
```

## estimate sPLSDA train and test error rates (Concateantion and Ensemble)

```{r}
# ## Concatenation
# modName = "Concatenation_sPLSDA"
# keepX <- splsdaConcat$choice.keepX
# X_train <- do.call(cbind, X.train[names(X.test)])
# X_test <- do.call(cbind, X.test)
# colnames(X_train) <- colnames(X_test) <- paste(rep(names(X.test), sapply(X.test, ncol)), unlist(sapply(X.test, colnames)), sep="_")
# concatSplsda_perf <- concatSplsdaFit(keepX, ncomp, X_train, X_test, y_train, y_test, folds, iter, modName)
# saveRDS(concatSplsda_perf, "~/Dropbox/PROOF/Manuscript/mixOmics/diablo_datasets/brca/concatSplsda_perf.rds")
concatSplsda_perf <- readRDS("~/Dropbox/PROOF/Manuscript/mixOmics/diablo_datasets/brca/concatSplsda_perf.rds")

# ## Ensemble
# modName = "Ensemble_sPLSDA"
# keepX <- lapply(tune_singleOmics_splsda, function(i) i$choice.keepX)
# all(names(keepX) == names(X_test))  # check order datasets
# X_train <- X.train; y_train <- Y.train; X_test <- X.test; y_test <- Y.test;
# ensembleSplsda_perf <- ensembleSplsdaFit(keepX, ncomp, X_train, X_test, y_train, y_test, folds, iter, modName)
# saveRDS(ensembleSplsda_perf, "~/Dropbox/PROOF/Manuscript/mixOmics/diablo_datasets/brca/ensembleSplsda_perf.rds")
ensembleSplsda_perf <- readRDS("~/Dropbox/PROOF/Manuscript/mixOmics/diablo_datasets/brca/ensembleSplsda_perf.rds")

```

## tune and estimate Enet train and test error rates (Concateantion and Ensemble)

```{r}
# penalties <- seq(0.01, 0.1, length.out = 10)
# modName <- "Concatenation_Enet";
# X_train=do.call(cbind, X.train[names(X.test)]);
# X_test=do.call(cbind, X.test)
# colnames(X_train) <- colnames(X_test) <- paste(rep(names(X.test), sapply(X.test, ncol)), colnames(X_test), sep="_")
# y_train=Y.train
# y_test=Y.test
# # Concatenation
# concatEnet_perf <- concatEnetFit(X_train, X_test, y_train, y_test, penalties, folds, iter, modName)
# saveRDS(concatEnet_perf, "~/Dropbox/PROOF/Manuscript/mixOmics/diablo_datasets/brca/concatEnet_perf.rds")
concatEnet_perf <- readRDS("~/Dropbox/PROOF/Manuscript/mixOmics/diablo_datasets/brca/concatEnet_perf.rds")

# ## Ensemble
# modName <- "Ensemble_Enet";
# X_train <- X.train[names(X.test)]
# X_test <- X.test
# ensembleEnet_perf <- ensembleEnetFit(X_train, X_test, y_train, y_test, penalties, folds, iter, modName)
# saveRDS(ensembleEnet_perf, "~/Dropbox/PROOF/Manuscript/mixOmics/diablo_datasets/brca/ensembleEnet_perf.rds")
ensembleEnet_perf <- readRDS("~/Dropbox/PROOF/Manuscript/mixOmics/diablo_datasets/brca/ensembleEnet_perf.rds")

```

## plot of error rates

```{r perf_comparison, fig.path='Figures/', dev='png', fig.height = 6, fig.width = 10}
rbind(diablonull_perf$error, diablofull_perf$error, concatSplsda_perf$error, ensembleSplsda_perf$error, concatEnet_perf$error, ensembleEnet_perf$error) %>% 
  mutate(cohort = gsub("train", "train (5x5-fold CV)", cohort)) %>% 
  mutate(cohort = gsub("test", "test (off-the-self)", cohort)) %>% 
  ggplot(aes(x = modName, y = mean_err, fill = cohort, color = cohort)) +
  theme(panel.grid.minor = element_line(colour="grey", size=0.5)) +
  geom_errorbar(aes(ymin = mean_err-sd_err, ymax = mean_err+sd_err), width=0.5, position = position_dodge(width = 0.5)) +
  geom_point(size=2, position = position_dodge(width = 0.5)) +
  ylim(c(0, 0.5)) + 
  customTheme(sizeStripFont = 15, xAngle = 30, hjust = 1, vjust = 1, xSize = 10, ySize = 10, xAxisSize = 10, yAxisSize = 10) +
  ylab("Balanced Error Rate (BER)") +
  xlab("Integrative methods")
```

## table of error rates

```{r}
errRates <- rbind(diablonull_perf$error, diablofull_perf$error, concatSplsda_perf$error, ensembleSplsda_perf$error, concatEnet_perf$error, ensembleEnet_perf$error) %>% 
  mutate(mean_err = signif(mean_err, 2),
    sd_err = signif(sd_err, 2))
kable(errRates)

write.csv(errRates,
  "~/Dropbox/PROOF/Manuscript/mixOmics/diablo_datasets/brca/brcaMethodsErrorRates.csv")
```

## numberof variables in each panel

```{r}
panelFeatures <- rbind(diablonull_perf$panels, diablofull_perf$panels, concatSplsda_perf$panels, ensembleSplsda_perf$panels, concatEnet_perf$panels, ensembleEnet_perf$panels)
kable(panelFeatures %>% spread(dataset, panelLength))

write.csv(panelFeatures %>% spread(dataset, panelLength),
  "~/Dropbox/PROOF/Manuscript/mixOmics/diablo_datasets/brca/brcaMethodsPanelSizes.csv")
```

# Analysis 2: DIABLO identified known and novel multi-omics biomarkers of breast cancer subtypes - using mRNA, miRNA, CpGs and proteins

We next demonstrate that DIABLO can identify novel biomarkers in addition to biomarkers with known biological associations using a case study of human breast cancer. We applied our biomarker analysis workflow to breast cancer datasets to characterize and predict PAM50 breast cancer subtypes. 

```{r out.width="100%"}
img1_path <- "/Users/asingh/Dropbox/PROOF/Manuscript/mixOmics/diablo/analyses/casestudy1_brca/Figures/SuppFig7.png"
include_graphics(img1_path)
```

*A standard DIABLO workflow. The first step inputs multiple omics datasets measured on the same individuals, that were previously normalized and filtered, , along with the phenotype information indicating the class membership of each sample (two or more groups). Optional preprocessing steps include multilevel transformation for repeated measures study designs and pathway module summary transformations. DIABLO is a multivariate dimension reduction method that seeks for latent components – linear combinations of variables from each omics dataset, that are maximally correlated as specified by a design matrix (see Methods section). The identification of a multi-omics panel is obtained with l1 penalties in the model that shrink the variable coefficients defining the components to zero. Numerous visualizations are proposed to provide insights into the multi-omics panel and guide the interpretation of the selected omics variables, including sample and variable plots. Downstream analysis include gene set enrichment analysis.*

After preprocessing and normalization of each omics data-type, the samples were divided into training and test sets.

```{r out.width="100%"}
img1_path <- "/Users/asingh/Dropbox/PROOF/Manuscript/mixOmics/diablo/analyses/casestudy1_brca/Figures/Table1.png"
include_graphics(img1_path)
```

*Overview of multi-omics datasets analyzed for method benchmarking and in two case studies. The breast cancer case study includes training and test datasets for all omics types except proteins.*

The training data consisted of four omics-datasets (mRNA, miRNA, CpGs and proteins) whereas the test data included all remaining samples for which the protein expression data were missing. The optimal multi-omics biomarker panel size was identified using a grid approach where, for any given combination of variables, we assessed the classification performance using a 5-fold cross-validation repeated 5 times. 

```{r out.width="100%"}
img1_path <- "/Users/asingh/Dropbox/PROOF/Manuscript/mixOmics/diablo/analyses/casestudy1_brca/Figures/optimal_errorRate_tuneFunction_mixOmics-1.png"
include_graphics(img1_path)
```

*Breast cancer multi omics study: optimal multi-omics biomarker panel for PAM50 subtypes. A grid was used to identify the optimal combination of variables select from each omics datasets. The following grid values was used for each omics dataset: mRNA = [5, 10, 15, 20], miRNA = [5, 10, 15, 20], CpGs = [5, 10, 15, 20], Proteins = [5, 10, 15, 20], across 3 components. The centroids distance measure was used to compute the error rate. The optimal multi-omics panel consisted of 20 mRNAs, 20 miRNAs, 15 CpGs and 15 proteins on component 1, 5 mRNAs, 5 miRNAs, 5 CpGs and 20 proteins on component 2, and 20 mRNAs, 20 miRNAs, 5 CpGs and 20 proteins on component 3.*

## tuning a DIABLO_full model using 4 datasets

```{r optimal_errorRate_tuneFunction_mixOmics, fig.path='Figures/', dev='png', fig.height = 10, fig.width = 20}
## design matrix
design <- matrix(1, nrow = length(X.train), ncol = length(X.train))
rownames(design) <- colnames(design) <- names(X.train)
diag(design) <- 0
# test.keepX = list(mRNA = c(5, 10, 15, 20), miRNA = c(5, 10, 15, 20), CpGs = c(5, 10, 15, 20), Proteins = c(5, 10, 15, 20))
# 
# t1 <- proc.time()
# tune = tune.block.splsda(X = X.train, Y = Y.train, ncomp = 3, test.keepX = test.keepX, dist = "centroids.dist",
#                           design = design, validation = "Mfold", folds = 5, nrepeat = 5, cpus = 5)
# t2 <- proc.time()
# (t2-t1)/60 #  minutes
# saveRDS(tune, "~/Dropbox/PROOF/Manuscript/mixOmics/diablo_datasets/brca/brcaPanel_tune_caseStudy.rds")

tune <- readRDS("~/Dropbox/PROOF/Manuscript/mixOmics/diablo_datasets/brca/brcaPanel_tune_caseStudy.rds")

## Component 1
p1 <- data.frame(mean = tune$error.rate[, "comp1"], sd = tune$error.rate.sd[, "comp1"]) %>% 
  as.data.frame %>% 
  mutate(keep = rownames(.)) %>% 
  arrange(mean) %>% 
  mutate(keep = factor(keep, keep), Comp = "Component 1") %>% 
  slice(1:50) %>% 
  ggplot(aes(y = keep, x = mean)) +
  geom_point() +
  geom_errorbarh(aes(xmin = mean-sd,  xmax = mean+sd)) +
  customTheme(sizeStripFont = 25, xAngle = 0, hjust = 0.5, vjust = 0.5, xSize = 15, 
              ySize = 15, xAxisSize = 15, yAxisSize = 15) +
  xlab("Mean±SD Error rate \n (5 x 5-fold cross-validation)") +
  ylab("Number of features per dataset (mRNA_miRNA_CpGs_Proteins) \n top 50 keepX with the minimum error rate for component 1") +
  facet_wrap(~Comp) +
  annotate("text", label = "minError = 31.5% \n (20_15_10_20)", x = 0.38, y = 0.75, size = 5) +
  xlim(c(0.28,0.4))

## Component 2
p2 <- data.frame(mean = tune$error.rate[, "comp2"], sd = tune$error.rate.sd[, "comp2"]) %>% 
  as.data.frame %>% 
  mutate(keep = rownames(.)) %>% 
  arrange(mean) %>% 
  mutate(keep = factor(keep, keep), Comp = "Component 2") %>% 
  slice(1:50) %>% 
  ggplot(aes(y = keep, x = mean)) +
  geom_point() +
  geom_errorbarh(aes(xmin = mean-sd,  xmax = mean+sd)) +
  customTheme(sizeStripFont = 25, xAngle = 0, hjust = 0.5, vjust = 0.5, xSize = 15, 
              ySize = 15, xAxisSize = 15, yAxisSize = 15) +
  xlab("Mean±SD Error rate \n (5 x 5-fold cross-validation)") +
  ylab("Number of features per dataset (mRNA_miRNA_CpGs_Proteins) \n top 50 keepX with the minimum error rate for component 2") +
  facet_wrap(~Comp) +
  annotate("text", label = "minError = 21.9% \n (20_15_10_20)", x = 0.3, y = 0.75, size = 5) +
  xlim(c(0.16,0.4))

## Component 3
p3 <- data.frame(mean = tune$error.rate[, "comp3"], sd = tune$error.rate.sd[, "comp3"]) %>% 
  as.data.frame %>% 
  mutate(keep = rownames(.)) %>% 
  arrange(mean) %>% 
  mutate(keep = factor(keep, keep), Comp = "Component 3") %>% 
  slice(1:50) %>% 
  ggplot(aes(y = keep, x = mean)) +
  geom_point() +
  geom_errorbarh(aes(xmin = mean-sd,  xmax = mean+sd)) +
  customTheme(sizeStripFont = 25, xAngle = 0, hjust = 0.5, vjust = 0.5, xSize = 15, 
              ySize = 15, xAxisSize = 15, yAxisSize = 15) +
  xlab("Mean±SD Error rate \n (5 x 5-fold cross-validation)") +
  ylab("Number of features per dataset (mRNA_miRNA_CpGs_Proteins) \n top 50 keepX with the minimum error rate for component 3") +
  facet_wrap(~Comp) +
  annotate("text", label = "minError = 17.9% \n (20_15_10_20)", x = 0.25, y = 0.75, size = 5) +
  xlim(c(0.15,0.3))

plot_grid(p1, p2, p3, ncol = 3)
```

## Optimal DIABLO model

The optimal multi-omics panel consisted of 45 mRNA, 45 miRNAs, 25 CpGs and 55 proteins selected across three components with a balanced error rate of 17.9±1.9%. 

```{r, results='hide'}
## check rownames ordering
all(rownames(tune$error.rate) == rownames(tune$error.rate.sd))
```

```{r}
data.frame(keepX = rep(rownames(tune$error.rate), 3), 
  meanError = as.numeric(tune$error.rate), 
  sdError = as.numeric(tune$error.rate.sd),
  comp = rep(paste0("comp", 1:3), each=nrow(tune$error.rate))) %>% 
  group_by(comp) %>% 
  filter(meanError == min(meanError))
```

### optimal keepX

```{r}
tune$choice.keepX
```

## run DIABLO - with optimal keepX

```{r}
## design matrix
design <- matrix(1, nrow = length(X.train), ncol = length(X.train))
rownames(design) <- colnames(design) <- names(X.train)
diag(design) <- 0
ncomp <- 3
keepX = tune$choice.keepX

diablo = block.splsda(X = X.train, Y = Y.train, ncomp = ncomp, keepX = keepX, design = design)
```

### Number of variables of each omic-type in the diablo panel

```{r}
diabloPanel <- mapply(function(x, y, z){
  c(x, y, z)
}, x = lapply(selectVar(diablo, comp = 1)[-5], function(i) unlist(i[[1]])),
   y = lapply(selectVar(diablo, comp = 2)[-5], function(i) unlist(i[[1]])),
   z = lapply(selectVar(diablo, comp = 3)[-5], function(i) unlist(i[[1]])))
sapply(diabloPanel, length)

## only keep unique features
diabloPanel <- lapply(diabloPanel, unique)
```

### overlap between the different omic compartments (mRNA,miRNA,CpGs and Protein)
  * all mRNA, CpGs and proteins have been converted to gene symbols

```{r}
venn(diabloPanel)
```

#### overlap between the mRNA and CpGs

```{r}
intersect(diabloPanel$mRNA, diabloPanel$CpGs)
```

#### overlap between the mRNA and Proteins

```{r}
intersect(diabloPanel$mRNA, diabloPanel$Proteins)
```

### overlap between the diablo panel features (mRNA,miRNA,CpGs and Protein) and with curated databases

```{r}
driverdbv2 <- read.delim(paste0(WhereAmI, "disease_collections/driverdbv2.txt"), header = FALSE) %>% 
  as.matrix %>% 
  as.character()

mircancer <- read.delim(paste0(WhereAmI, "disease_collections/miRCancerOctober2017.txt"), header = FALSE)
mircancer <- mircancer[grep("breast", tolower(as.character(mircancer$V2))), "V1"] %>% as.character() %>% unique()

molsigdb <- collections[grep("BREAST", collections$geneset), "members_mrna"] %>% unlist %>% unique

omim0 <- read.delim(paste0(WhereAmI, "disease_collections/OMIM_Disease.txt"), header = FALSE)
omim <- apply(omim0[, -1], 1, function(i){
    x <- as.character(as.matrix(i))[as.character(as.matrix(i)) != ""]
    x[!is.na(x)]
  })
names(omim) <- as.character(omim0[,1])
omim <- omim[grep("breast", names(omim), value = TRUE)] %>% 
  unlist %>% unique

brca_genes = list(DIABLO = unlist(unlist(diabloPanel)),
                  MolSigDB = molsigdb,
                  DriverDBv2 = driverdbv2,
                  OMIM = omim,
  miRCancer=mircancer)
venn(brca_genes, ilab=TRUE, zcolor = "style")


brcaDatabase <- data.frame(database = rep(c("MolSigDB", "DriverDBv2", "OMIM", "miRCancer"), c(length(molsigdb), length(driverdbv2), length(omim), length(mircancer))),
  feature = c(molsigdb, driverdbv2, omim, mircancer))

```

## Feature Plot

This panel identified many variables with previously known associations with breast cancer, as assessed by looking at the overlap between the panel features and gene sets related to breast cancer based on the Molecular Signature database (MolSigDB) [23], miRCancer [24], Online Mendelian Inheritance in Man (OMIM) [25], and DriverDBv2 [26]. The feature plot depicts the variable contributions of each omics-type indicated by their loading weight (variable importance). Variables not found in any database may represent novel biomarkers of breast cancer.

```{r brcaPanel_features, fig.path='Figures/', dev='png', fig.width = 10, fig.height=13}
databases <- c("No overlap", "miRCancer", "DriverDBv2", "MolSigDB", "MolSigDB_DriverDBv2", "MolSigDB_DriverDBv2_OMIM", "MolSigDB_OMIM")
featurePlot = function(panel, omicName, database){
  Dat <- rbind(selectVar(diablo, comp = 1)[[omicName]][["value"]], 
  selectVar(diablo, comp = 2)[[omicName]][["value"]],
  selectVar(diablo, comp = 3)[[omicName]][["value"]]) %>% 
  data.frame() %>% 
  mutate(dataset = omicName,
         feature = gsub(";", "\n", rownames(.)),
         value.var = abs(value.var)) %>% 
  group_by(dataset) %>% 
  arrange(value.var) %>% 
  mutate(feature = factor(as.character(feature), feature))
  Dat$database <- sapply(as.character(Dat$feature), function(i){
  paste(unique(as.character(subset(brcaDatabase, feature %in% unlist(strsplit(i, ";")))$database)), collapse = "_")})
  Dat$database[Dat$database == ""] <- "No overlap"
  Dat$database <- factor(as.character(Dat$database), databases)
  
  Segment <- geom_segment(aes(yend = feature, color = database), xend = 0, size = 1)
  Point <- geom_point(size = 4, aes(color = database, fill = database)) 
  Theme <- theme_bw() + theme(panel.grid.major.y = element_blank(),legend.justification = c(1, 0.5))
  
  ggplot(Dat, aes(x = value.var, y = feature)) +
 	Segment +
 	Point +
	Theme +
	ggtitle(omicName) + ylab("Variables")
}

p1 <- featurePlot(panel=diabloPanel, omicName="mRNA", database=database) +
  theme(legend.position = c(0.95,0.15)) + scale_color_manual(values=c("#000000", "#E69F00", "#56B4E9")) + 
  theme(legend.title = element_text(size=16), 
    legend.text = element_text(size=16), 
    plot.title = element_text(size=26, face = "bold")) +
  customTheme(sizeStripFont = 25, xAngle = 0.5, hjust = 0.5, vjust = 0.5,
              xSize = 15, ySize = 9, xAxisSize = 20, yAxisSize = 20) + 
  scale_x_continuous("", limits = c(0, 1), breaks = seq(0,1,0.2)) 
p2 <- featurePlot(panel=diabloPanel, omicName="miRNA", database=database) +
  theme(legend.position = c(0.95,0.15)) + ylab("") + scale_color_manual(values=c("#000000", "#0072B2")) + 
  theme(legend.title = element_text(size=16), 
    legend.text = element_text(size=16), 
    plot.title = element_text(size=26, face = "bold")) +
  customTheme(sizeStripFont = 25, xAngle = 0.5, hjust = 0.5, vjust = 0.5,
              xSize = 15, ySize = 9, xAxisSize = 20, yAxisSize = 20) + 
  scale_x_continuous("", limits = c(0, 1), breaks = seq(0,1,0.2))
p3 <- featurePlot(panel=diabloPanel, omicName="CpGs", database=database) +
  theme(legend.position = c(0.95,0.15)) + scale_color_manual(values=c("#000000", "#E69F00")) + 
  theme(legend.title = element_text(size=16), 
    legend.text = element_text(size=16), 
    plot.title = element_text(size=26, face = "bold")) +
  customTheme(sizeStripFont = 25, xAngle = 0.5, hjust = 0.5, vjust = 0.5,
              xSize = 15, ySize = 9, xAxisSize = 20, yAxisSize = 20) + 
  scale_x_continuous("Variable contributions \n absolute loadings", limits = c(0, 1), breaks = seq(0,1,0.2))
p4 <- featurePlot(panel=diabloPanel, omicName="Proteins", database=database) +
  theme(legend.position = c(0.95,0.2)) + ylab("") + scale_color_manual(values=c("#000000", "#CC79A7", "#E69F00", "#56B4E9", "#009E73", "#F0E442")) + 
  theme(legend.title = element_text(size=16), 
    legend.text = element_text(size=10), 
    plot.title = element_text(size=26, face = "bold")) +
  customTheme(sizeStripFont = 25, xAngle = 0.5, hjust = 0.5, vjust = 0.5,
              xSize = 15, ySize = 9, xAxisSize = 20, yAxisSize = 20) + 
  scale_x_continuous("Variable contributions \n absolute loadings", limits = c(0, 1), breaks = seq(0,1,0.2))

plot_grid(p1, p2, p3, p4, ncol = 2)
```

## Evaluate performance of diablo panel using additional data (test datasets)

```{r, results='hide'}
## Training data
Y.train <- droplevels(pam50Train0$Call)
names(Y.train) <- rownames(pam50Train0)
X.train <- list(mRNA = mrnaTrain0, miRNA = mirnaTrain0, CpGs = methTrain0, Proteins = protTrain0)
all(names(Y.train) == rownames(X.train[[1]]))
all(names(Y.train) == rownames(X.train[[2]]))
all(names(Y.train) == rownames(X.train[[3]]))
all(names(Y.train) == rownames(X.train[[4]]))
dim(X.train[[1]]); dim(X.train[[2]]); dim(X.train[[3]]); dim(X.train[[4]]);
## Test data
Y.test <- droplevels(pam50Test0$Call)
names(Y.test) <- rownames(pam50Test0)
length(c(Y.train, Y.test))
X.test <- list(mRNA = mrnaTest0, miRNA = mirnaTest0, CpGs = methTest0)
all(names(Y.test) == rownames(X.test[[1]]))
all(names(Y.test) == rownames(X.test[[2]]))
all(names(Y.test) == rownames(X.test[[3]]))
dim(X.test[[1]]); dim(X.test[[2]]); dim(X.test[[3]]); 
```

### Number of samples in the train and test datasets

```{r}
rbind(table(Y.train), table(Y.test)) %>% 
  as.data.frame %>% 
  mutate(Set = c("Train", "Test"))
```

## Individual class error rate per PAM50 subtype

```{r}
## build diablo model and test on independent dataset
pred <- predict(diablo, X.test, dist = "all")

TestError <- apply(pred$WeightedVote$centroids.dist, 2, function(y) {
      y[is.na(y)] <- nlevels(Y.test) + 5
      temp = table(factor(y, levels = c(levels(Y.test), 
        nlevels(Y.test) + 5)), Y.test)
      diag(temp) <- 0
      err = c(colSums(temp)/summary(Y.test), sum(temp)/length(Y.test), 
        mean(colSums(temp)/summary(Y.test)))
      return(err = err)
    })
rownames(TestError) <- c(levels(Y.test), "Overall.ER", "Overall.BER")
TestError[, "comp3"]
```

## Component plots

The plot below depicts the consensus and individual omics component plots based on the optimal biomarker panel, along with 95% confidence ellipses obtained from the training data and superimposed with the samples from the test data. The majority of the samples were within the ellipses, suggesting a reproducible multi-omics biomarker panel from the training to the test set, that was predictive of breast cancer subtypes (balanced error rate = 22.9%). The consensus plot corresponded strongly with the mRNA component plot, depicting a strong separation of the Basal (error rate = 4.9%) and Her2 (error rate = 20%) subtypes. We observed a weak separation of Luminal A (LumA, error rate = 13.3%) and Luminal B (LumB, error rate = 53.3%) subtypes. 

```{r SamplePlots_trainEllipses_testPoints, fig.path='Figures/', dev='png', fig.height=13, fig.width=10}
# model variates
varDat.train <- do.call(rbind, diablo$variates[1:length(X.train)]) %>% 
  as.data.frame %>% mutate(subj = rownames(.))
varDat.train$Dataset <- rep(names(X.train), e = length(Y.train))
varDat.train$Class <- Y.train
varDat.train <- varDat.train %>% group_by(Class, subj) %>% 
  summarise_all(funs(mean)) %>% 
  mutate(Dataset = "Consensus") %>% 
  dplyr::select(`comp 1`, `comp 2`, `comp 3`, subj, Dataset, Class) %>% 
  as.data.frame() %>% 
  rbind(., varDat.train) %>% 
  mutate(Dataset = factor(Dataset, c("Consensus", "mRNA", "miRNA", "CpGs", "Proteins"))) 

# test variates
varDat.test <- do.call(rbind, pred$variates) %>% as.data.frame %>% mutate(subj = rownames(.))
varDat.test$Dataset <- rep(names(pred$variates), e = nrow(pred$variates[[1]]))
varDat.test$Class <- Y.test
colnames(varDat.train) <- colnames(varDat.test) <- c("comp 1", "comp 2", "comp 3", "subj", "Dataset", "Class")
varDat.test <- varDat.test %>% group_by(Class, subj) %>% 
  summarise_all(funs(mean)) %>% 
  mutate(Dataset = "Consensus") %>% 
  dplyr::select(`comp 1`, `comp 2`, `comp 3`, subj, Dataset, Class) %>% 
  as.data.frame() %>% 
  rbind(., varDat.test) %>% 
  mutate(Dataset = factor(Dataset, c("Consensus", "mRNA", "miRNA", "CpGs", "Proteins"))) 

p1 <- filter(varDat.test, Dataset == "Consensus") %>% 
  ggplot(aes(x = `comp 1`, y = `comp 2`, color = Class)) + 
  geom_point(size = 2) +
  facet_wrap(~Dataset, scales = "free", ncol = 5) + 
   stat_ellipse(data = filter(varDat.train, Dataset == "Consensus"), size = 1) +
  customTheme(sizeStripFont = 25, xAngle = 0.5, hjust = 0.5, vjust = 0.5,
              xSize = 20, ySize = 20, xAxisSize = 20, yAxisSize = 20) +
  xlab("Component 1") + ylab("Component 2") + 
  theme(strip.text.x = element_text(size=26, face = "bold")) + 
  scale_color_manual(values=color.mixo(1:4)) +
  annotate("text", x = -1.5, y = -3.5, label = "Ellipses: training data \n points: test data", size =8)

p2 <- filter(varDat.test, Dataset != "Consensus") %>% 
  ggplot(aes(x = `comp 1`, y = `comp 2`, color = Class)) + 
  geom_point(size = 2) +
  facet_wrap(~Dataset, scales = "free", ncol = 2) + 
   stat_ellipse(data = filter(varDat.train, Dataset != "Consensus"), size = 1) +
  customTheme(sizeStripFont = 25, xAngle = 0.5, hjust = 0.5, vjust = 0.5,
              xSize = 20, ySize = 20, xAxisSize = 20, yAxisSize = 20) +
  xlab("Component 1") + ylab("Component 2") + 
  theme(strip.text.x = element_text(size=26, face = "bold")) + 
  scale_color_manual(values=color.mixo(1:4))

plot_grid(p1, p2, ncol=1) 
```

## Heatmap

Similarly, the heatmap showing the scaled expression of all features of the multi-omics biomarker panel, depicted a strong clustering of the Basal and Her2 samples whereas the Luminal A and B were mixed. 

```{r heatmap_brca, fig.path='Figures/', dev='png', fig.height=10, fig.width=8}
cimDiablo(diablo, row.names = FALSE, col.names = FALSE)
```

## Network

Overall, the features of the multi-omics biomarker panel formed a densely connected network comprising of four communities where variables in each community (cluster) were densely connected with themselves and sparsely connected with other clusters. 

```{r network_brca, fig.path='Figures/', dev='png', fig.height=8, fig.width=8}
corMat <- mapply(function(x, y){
  y[, x]
}, x = diabloPanel, y = X.train, SIMPLIFY = FALSE) %>% 
  do.call(cbind, .) %>% 
  cor
colnames(corMat) <- rownames(corMat) <- paste(rep(names(diabloPanel), sapply(diabloPanel, length)), colnames(corMat), sep="_")

corMat[lower.tri(corMat)] <- 0
diag(corMat) <- 0

color.blocks = brewer.pal(n = 12, name = "Paired")[seq(2, 12, by = 2)]
links <- corMat %>% 
  as.data.frame() %>% 
  mutate(to = rownames(.)) %>% 
  gather(from, cor, -to) %>% 
  filter(abs(cor) > 0.4) %>% 
  mutate(Color = ifelse(cor > 0, "red", "blue"))
nodes <- data.frame(id = unique(c(links$to, links$from)))
nodes$datasets <- sapply(strsplit(as.character(nodes$id), "_"), function(i) i[1])
net <- graph_from_data_frame(d=links, vertices=nodes, directed=FALSE) 
E(net)$color <- links$Color
V(net)$color <- color.blocks[as.numeric(factor(nodes$datasets, names(diabloPanel)))]
E(net)$weight <- abs(links$cor)
weight <- E(net)$weight
weight[weight == 0.5] <- 0.3

ceb <- cluster_edge_betweenness(net) 
mark.groups <- split(as.character(nodes$id), factor(ceb$membership))
mark.groups <- mark.groups[sapply(mark.groups, length) > 3]

mark.col <- rainbow(length(mark.groups), alpha=.5)
plot(net, edge.curved=.2, vertex.label.cex=0.01, vertex.size=6, mark.groups=mark.groups, 
  mark.col=mark.col, vertex.label.color="black") 
legend("bottomleft", names(diabloPanel), col=unique(V(net)$color), pch = 19, bty="n", cex = 1.5)
```

### Number of variables of each omic-type in the red cluster

```{r}
sapply(names(diabloPanel), function(i){
  length(grep(i, mark.groups$`1`))
})
```

## Geneset enrichment analysis

The largest cluster in the network consisted of 72 variables; 20 mRNAs, 21 miRNAs, 15 CpGs and 16 proteins (red bubble) and was further investigated using gene set enrichment analysis. We identified many cancer-associated pathways (e.g. FOXM1 pathway, p53 signaling pathway), DNA damage and repair pathways (e.g. E2F mediated regulation of DNA replication, G2M DNA damage checkpoint) and various cell-cycle pathways (e.g. G1S transition, mitotic G1/G1S phases), demonstrating the ability of DIABLO to identify a biologically plausible multi-omics biomarker panel. This panel generalized to new breast cancer samples and implicated previously unknown molecular features in breast cancer, which could be further validated in experimental studies. 

```{r genesetEnrichment_brca, fig.path='Figures/', dev='png', fig.height=7, fig.width=7}
## perform gene set enrichment analysis on each cluster of features
## intersection > 5 genes
input <- unlist(strsplit(sapply(strsplit(mark.groups[[1]], "_"), function(i) i[2]), ";"))
enrichment <- sear(input, "mrna") %>% 
    group_by(collection, subcollection) %>% 
    filter(fdr < 0.05, collection %in% "C2") %>% 
    filter(subcollection != "CGP")

enrichment %>% ungroup %>% 
  mutate(geneset = tolower(sapply(strsplit(as.character(geneset), "_"), function(i) paste(i[-1], collapse=" ")))) %>% 
    arrange(desc(fdr)) %>% 
    mutate(geneset = factor(geneset, unique(geneset))) %>% 
  ggplot(aes(x = geneset, y = -log10(fdr), fill = subcollection, color = subcollection)) +
  geom_bar(stat = "identity", position = "dodge") +
  coord_flip() +
  customTheme(sizeStripFont = 15, xAngle = 0, hjust = 0.5, vjust = 0.5, 
    xSize = 20, ySize = 9, xAxisSize = 20, yAxisSize = 20) +
  theme(legend.position = c(0.45,0.1)) +
  ylab("Gene sets") +
  xlab("Signficant Geneset at FDR=5%")

```

## References
3. Wang B, Mezlini AM, Demir F, Fiume M, Tu Z, Brudno M, et al. Similarity network fusion for aggregating data types on a genomic scale. Nat Methods [Internet]. 2014 [cited 2016 Jan 19];11:333–7. Available from: http://www.nature.com/doifinder/10.1038/nmeth.2810
12. Rohart F, Gautier B, Singh A, Cao K-AL. mixOmics: An R package for ‘omics feature selection and multiple data integration. PLOS Comput Biol [Internet]. 2017 [cited 2018 Jan 29];13:e1005752. Available from: http://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1005752
19. The TCGA Research Network. The Cancer Genome Atlas [Internet]. Available from: http://cancergenome.nih.gov/
20. Singh A, Yamamoto M, Kam SHY, Ruan J, Gauvreau GM, O’Byrne PM, et al. Gene-metabolite expression in blood can discriminate allergen-induced isolated early from dual asthmatic responses. Hsu Y-H, editor. PLoS ONE [Internet]. 2013 [cited 2015 Jul 18];8:e67907. Available from: http://dx.plos.org/10.1371/journal.pone.0067907
21. Singh A, Yamamoto M, Ruan J, Choi JY, Gauvreau GM, Olek S, et al. Th17/Treg ratio derived using DNA methylation analysis is associated with the late phase asthmatic response. Allergy Asthma Clin Immunol [Internet]. 2014 [cited 2016 Mar 2];10:32. Available from: http://www.biomedcentral.com/content/pdf/1710-1492-10-32.pdf
23. Liberzon A, Birger C, Thorvaldsdóttir H, Ghandi M, Mesirov JP, Tamayo P. The Molecular Signatures Database Hallmark Gene Set Collection. Cell Syst [Internet]. 2015 [cited 2018 Jan 30];1:417–25. Available from: http://linkinghub.elsevier.com/retrieve/pii/S2405471215002185
24. Xie B, Ding Q, Han H, Wu D. miRCancer: a microRNA-cancer association database constructed by text mining on literature. Bioinformatics [Internet]. 2013 [cited 2018 Jan 30];29:638–44. Available from: https://academic.oup.com/bioinformatics/article-lookup/doi/10.1093/bioinformatics/btt014
25. Hamosh A. Online Mendelian Inheritance in Man (OMIM), a knowledgebase of human genes and genetic disorders. Nucleic Acids Res [Internet]. 2004 [cited 2018 Jan 30];33:D514–7. Available from: https://academic.oup.com/nar/article-lookup/doi/10.1093/nar/gki033
26. Chung I-F, Chen C-Y, Su S-C, Li C-Y, Wu K-J, Wang H-W, et al. DriverDBv2: a database for human cancer driver gene research. Nucleic Acids Res [Internet]. 2016 [cited 2018 Jan 30];44:D975–9. Available from: https://academic.oup.com/nar/article-lookup/doi/10.1093/nar/gkv1314