---
title: "Benchmarking"
output:
  rmdformats::readthedown:
    self_contained: true
    thumbnails: true
    lightbox: true
    gallery: false
    highlight: tango
---

```{r setup, message=FALSE, warning=FALSE, tidy=TRUE, echo=FALSE}
knitr::opts_chunk$set(fig.pos = 'p', message=FALSE, warning=FALSE, tidy=TRUE, echo=FALSE) # Places figures on their own pages
knitr::opts_chunk$set(out.width = '100%', dpi=300)

WhereAmI <- "~/Dropbox/PROOF/Manuscript/mixOmics/diablo/analyses/benchmarking/"

## load libraries
library(knitr)
library(cowplot)
library(r.jive)
library(tidyverse)
library(MOFAtools); library(MultiAssayExperiment);
library(sear)
library(grid)
library(mixOmics)
library(amritr)
library(org.Hs.eg.db)
library(UpSetR)
library(venn)
library(ggalt)
library(ggrepel)
library(igraph)
library(ggraph)
source(paste0(WhereAmI, "functions_connectivity_biologicalEnrichment.R"))

## load data
load("~/Dropbox/PROOF/Manuscript/mixOmics/diablo_datasets/benchmarking/SNFdatasets.RDATA")
load("~/Dropbox/PROOF/Manuscript/mixOmics/diablo_datasets/benchmarking/jive/jive_SNFdatasets.RDATA")
load("~/Dropbox/PROOF/Manuscript/mixOmics/diablo_datasets/benchmarking/mofa/mofa_SNFdatasets.RDATA")

## global variables
superCol <- "#999999"
unsuperCol <- "#F0E442"
```

## DIABLO identifies molecular networks with superior biological enrichment

To assess this, we turn to real biological datasets. We applied various integrative approaches to cancer multi-omics datasets (mRNA, miRNA, and CpG) – colon, kidney, glioblastoma (gbm) and lung – and identified multi-omics biomarker panels that were predictive of high and low survival times. We then compared the network properties and biological enrichment of the selected features across approaches.

```{r out.width="100%"}
img1_path <- "/Users/asingh/Dropbox/PROOF/Manuscript/mixOmics/diablo/analyses/casestudy1_brca/Figures/Table1.png"
include_graphics(img1_path)
```

*Overview of multi-omics datasets analyzed for method benchmarking and in two case studies. The breast cancer case study includes training and test datasets for all omics types except proteins.*

## SNF data description
  * The SNF datasets were part of the datasets used in the Nature Methods paper on Similarity Network Fusion (SNF); https://www.nature.com/articles/nmeth.2810
  * The cancer datasets include GBM (Brain), Colon, Kidney, Lung and Breast (the Breast cancer dataset was excluded in order to avoid confusion with the case study on Breast Cancer)
  * The datasets were obtained from: http://compbio.cs.toronto.edu/SNF/SNF/Software.html
  * Survival times were provided for each disease cohort. The median survival time was used to dictomize each response variables into low and high survival times.

## number of samples in each group

```{r}
addmargins(sapply(snf_group, table))
```

## number of variables in each dataset
* mRNA transcripts or cpg probes that mapped to the same gene were averaged 

```{r}
sapply(snf_data, function(i) sapply(i, ncol))
```

# Multi-omic biomarker panels

Multi-omics biomarker panels were developed using component-based integrative approaches that also performed variable selection: supervised methods included concatenation and ensemble schemes using the sPLSDA classifier [14], and DIABLO with either the null or full design (DIABLO_null, and DIABLO_full); unsupervised approaches included sparse generalized canonical correlation analysis [15] (sGCCA), Multi-Omics Factor Analysis (MOFA), and Joint and Individual Variation Explained (JIVE) [23] (see Supplementary Note for parameter settings). Both supervised and unsupervised approaches were considered in order to compare and contrast the types of omics-variables selected, network properties and biological enrichment results. A distinction was made between DIABLO models in which the correlation between omics datasets was not maximized (DIABLO_null) and those when the correlation between omics datasets was maximized (DIABLO_full).

## Unsupervised

### JIVE

```{r}
jive_joints <- list(colon = t(do.call(rbind, jive_colon$joint)),
  kidney = t(do.call(rbind, jive_kidney$joint)),
  gbm = t(do.call(rbind, jive_gbm$joint)),
  lung = t(do.call(rbind, jive_lung$joint)))
colnames(jive_joints$colon) <- sapply(snf_data$colon, colnames) %>% unlist
colnames(jive_joints$kidney) <- sapply(snf_data$kidney, colnames) %>% unlist
colnames(jive_joints$gbm) <- sapply(snf_data$gbm, colnames) %>% unlist
colnames(jive_joints$lung) <- sapply(snf_data$lung, colnames) %>% unlist

snf_jive <- mapply(function(x, y){
  # run sPCA
  pca_jive = spca(x, ncomp = 2, center = FALSE, scale = FALSE, keepX = rep(ncol(x), 2))
  
  ## scores
  pcs <- pca_jive$variates$X %>% 
         as.data.frame %>% 
         mutate(pheno = y, Method = "JIVE")

  ## features 
  comp1 <- split(pca_jive$loadings$X[, 1], factor(sapply(strsplit(names(pca_jive$loadings$X[, 1]), "_"), function(i) i[1]), c("mrna", "mirna", "cpg")))
  comp2 <- split(pca_jive$loadings$X[, 2], factor(sapply(strsplit(names(pca_jive$loadings$X[, 2]), "_"), function(i) i[1]), c("mrna", "mirna", "cpg")))
  
  panel <- mapply(function(x, y){
    c(names(x[order(abs(x), decreasing = TRUE)][1:30]),
      names(y[order(abs(y), decreasing = TRUE)][1:30]))
  }, x = comp1, y = comp2, SIMPLIFY = FALSE)
  
  return(list(pcs=pcs, panel=panel))
}, x = jive_joints, y = snf_group, SIMPLIFY = FALSE) %>% zip_nPure()

```

### MOFA

```{r}
mofa_joints <- list(colon = MOFA_colon, kidney = MOFA_kidney, gbm = MOFA_gbm, lung = MOFA_lung)

snf_mofa <- mapply(function(x, y, z){
  ## scores
  pcs <- getFactors(x, factors = "all", as.data.frame = FALSE, include_intercept = TRUE) %>% as.data.frame()
  colnames(pcs) <- c("intercept", "PC1", "PC2")
  pcs <- pcs[, -1] %>% 
         mutate(pheno = y, Method = "MOFA")

  ## features
  features_mofa = getWeights(x, views = "all", factors = 1:2)
  features_mofa <- features_mofa[names(z)]
  panel <- mapply(function(x, y){
    c(colnames(y)[order(abs(x[, 1]), decreasing = TRUE)[1:30]],
      colnames(y)[order(abs(x[, 2]), decreasing = TRUE)[1:30]])
    }, x = features_mofa, y = z, SIMPLIFY = FALSE)
  
  return(list(pcs=pcs, panel=panel))
}, x = mofa_joints, y = snf_group, z = snf_data, SIMPLIFY = FALSE) %>% zip_nPure()

```

### sGCCA

```{r}
## design matrix
design <- matrix(1, nrow = 3, ncol = 3)
rownames(design) <- colnames(design) <- names(snf_data$colon)
diag(design) <- 0

keepX = lapply(snf_mofa$panel, function(i){
      lapply(i, function(i){
        rep(length(i)/2, 2)
      })
})

snf_sgcca <- mapply(function(x, y, z){
  result.unsupervised = wrapper.sgcca(X = x, ncomp = 2, keepX = z, design = design)
  
  ## scores
  pcs <- Reduce("+", result.unsupervised$variates)/2
  colnames(pcs) <- c("PC1", "PC2")
  pcs <- pcs %>% as.data.frame() %>% 
  mutate(pheno = y, Method = "sGCCA")

  ## features
  feat1 <- lapply(selectVar(result.unsupervised, comp = 1), function(i) i["name"])
  feat2 <- lapply(selectVar(result.unsupervised, comp = 2), function(i) i["name"])

  panel <- list(mrna = c(feat1$mrna$name, feat2$mrna$name),
                mirna = c(feat1$mirna$name, feat2$mirna$name),
                cpg = c(feat1$cpg$name, feat2$cpg$name))
  
  return(list(pcs=pcs, panel=panel))
}, x = snf_data, y = snf_group, z = keepX, SIMPLIFY = FALSE) %>% zip_nPure()


```

## Supervised

### Concatenation_sPLSDA

```{r}
snf_concat_splsda <- mapply(function(x, y){
  X = do.call(cbind, x)
  concat <- splsda(X = do.call(cbind, x), Y = y, ncomp = 2, keepX = rep(ncol(X), 2))
  ## scores
  pcs <- concat$variates$X %>% 
         as.data.frame
  colnames(pcs) <- c("PC1", "PC2")
  pcs <- pcs %>% mutate(pheno = y, Method = "Concatenation")
  
  ## features 
  comp1 <- split(concat$loadings$X[, 1], factor(sapply(strsplit(names(concat$loadings$X[, 1]), "_"), function(i) i[1]), c("mrna", "mirna", "cpg")))
  comp2 <- split(concat$loadings$X[, 2], factor(sapply(strsplit(names(concat$loadings$X[, 2]), "_"), function(i) i[1]), c("mrna", "mirna", "cpg")))
  
  panel <- mapply(function(x, y){
    c(names(x[order(abs(x), decreasing = TRUE)][1:30]), 
      names(y[order(abs(y), decreasing = TRUE)][1:30]))
  }, x = comp1, y = comp2, SIMPLIFY = FALSE)
  
  return(list(pcs=pcs, panel=panel))
}, x = snf_data, y = snf_group, SIMPLIFY = FALSE) %>% zip_nPure()


```

### Ensemble_spslda

```{r}
snf_ensemble_splsda <- mapply(function(x, y, z){
  ensem <- mapply(function(a, b){
     result <- splsda(X = a, Y = y, keepX = b, ncomp = 2)
  
     ## scores
     pcs <- result$variates$X %>% as.data.frame
     colnames(pcs) <- c("PC1", "PC2")
     
     ## panels
     feat <- c(selectVar(result, comp = 1)$name, selectVar(result, comp = 2)$name)
     
     return(list(pcs=pcs, feat=feat))
  }, a = x, b = z, SIMPLIFY = FALSE) %>% zip_nPure()
  
  ## scores
  pcs <- as.data.frame(Reduce("+", ensem$pcs)/3) %>% 
    mutate(pheno = y, Method = "Ensemble")
  
  ## features
  panel <- ensem$feat
  return(list(pcs = pcs, panel = panel))
}, x = snf_data, y = snf_group, z = keepX, SIMPLIFY = FALSE) %>% zip_nPure()


```

### DIABLO_null

```{r}
## design matrix
design <- matrix(0, nrow = 3, ncol = 3)
rownames(design) <- colnames(design) <- names(snf_data$colon)
diag(design) <- 0

snf_diabloNull <- mapply(function(x, y, z){
  result.supervised = block.splsda(X = x, Y = y, ncomp = 2, keepX = z, design = design)
  
  ## scores
  pcs <- Reduce("+", result.supervised$variates)/length(x)
  colnames(pcs) <- c("PC1", "PC2")
  pcs <- pcs %>% as.data.frame() %>% 
  mutate(pheno = y, Method = "DIABLO_null")

  ## features
  feat1 <- lapply(selectVar(result.supervised, comp = 1), function(i) i["name"])
  feat2 <- lapply(selectVar(result.supervised, comp = 2), function(i) i["name"])

  panel <- list(mrna = c(feat1$mrna$name, comp2 = feat2$mrna$name),
                mirna = c(feat1$mirna$name, comp2 = feat2$mirna$name),
                cpg = c(feat1$cpg$name, comp2 = feat2$cpg$name))
  
  return(list(pcs=pcs, panel=panel))
}, x = snf_data, y = snf_group, z = keepX, SIMPLIFY = FALSE) %>% zip_nPure()


```

### DIABLO_full

```{r}
## design matrix
design <- matrix(1, nrow = 3, ncol = 3)
rownames(design) <- colnames(design) <- names(snf_data$colon)
diag(design) <- 0

snf_diabloFull <- mapply(function(x, y, z){
  result.supervised = block.splsda(X = x, Y = y, ncomp = 2, keepX = z, design = design)
  
  ## scores
  pcs <- Reduce("+", result.supervised$variates)/length(x)
  colnames(pcs) <- c("PC1", "PC2")
  pcs <- pcs %>% as.data.frame() %>% 
  mutate(pheno = y, Method = "DIABLO_full")

  ## features
  feat1 <- lapply(selectVar(result.supervised, comp = 1), function(i) i["name"])
  feat2 <- lapply(selectVar(result.supervised, comp = 2), function(i) i["name"])

  panel <- list(mrna = c(feat1$mrna$name, comp2 = feat2$mrna$name),
                mirna = c(feat1$mirna$name, comp2 = feat2$mirna$name),
                cpg = c(feat1$cpg$name, comp2 = feat2$cpg$name))
  
  return(list(pcs=pcs, panel=panel))
}, x = snf_data, y = snf_group, z = keepX, SIMPLIFY = FALSE) %>% zip_nPure()


```

\pagebreak

# Number of features per panel

Each multi-omics biomarker panel included 180 features (60 features of each omics type across 2 components). Approaches generally identified distinct sets of features. The plots below depict the distinct and shared features between the seven multi-omics panels obtained from the unsupervised (purple, sGCCA, MOFA and JIVE) and supervised (green, Concatenation, Ensemble, DIABLO_null and DIABLO_full) methods. Supervised methods selected many of the same features (blue), but DIABLO_full had greater feature overlap with unsupervised methods (orange). 

```{r panelSizes, fig.path='Figures/', dev='png', fig.height = 5, fig.width = 5}
multiOmicPanels <- list(JIVE = snf_jive$panel, MOFA = snf_mofa$panel, sGCCA = snf_sgcca$panel,
                  Concatenation = snf_concat_splsda$panel, Ensemble = snf_ensemble_splsda$panel, 
                  DIABLO_null = snf_diabloNull$panel, DIABLO_full = snf_diabloFull$panel)

allPanels <- list(JIVE = snf_jive$panel, MOFA = snf_mofa$panel, sGCCA = snf_sgcca$panel,
                  Concatenation = snf_concat_splsda$panel, Ensemble = snf_ensemble_splsda$panel, 
                  DIABLO_null = snf_diabloNull$panel, DIABLO_full = snf_diabloFull$panel) %>% rapply(., function(i){
                    sapply(strsplit(as.character(i), "_"), function(j) as.character(j[[2]]))
                  }, how = "list")

panels <- rapply(multiOmicPanels, function(i){
  length(unique(i))
}, how = "list") %>% 
  lapply(., function(j){
  do.call(rbind, j) %>% 
    as.data.frame() %>% 
    mutate(disease=rownames(.))
}) %>% 
  do.call(rbind, .) %>% 
  as.data.frame() %>% 
  mutate(method = sapply(strsplit(rownames(.),"\\."), function(i) i[1])) %>% 
  gather(omic, nFeat, -c(disease:method)) %>% 
  mutate(nFeat=as.numeric(nFeat)) %>% 
  ggplot(aes(x = method, y = nFeat, fill = omic)) + 
  geom_bar(stat = "identity") +
  facet_grid(~disease) +
  customTheme(sizeStripFont = 10, xAngle = 45, hjust = 1, vjust = 1, 
              xSize = 10, ySize = 10, xAxisSize = 10, yAxisSize = 10)
panels

## single omic panels
mrna <- lapply(allPanels, function(i){
  lapply(i, function(j){
    j[["mrna"]]
  })
})
mirna <- lapply(allPanels, function(i){
  lapply(i, function(j){
    j[["mirna"]]
  })
})
cpg <- lapply(allPanels, function(i){
  lapply(i, function(j){
    j[["cpg"]]
  })
})

mrna.cpg <- lapply(names(mrna), function(i){
  x <- lapply(names(mrna$JIVE), function(j){
      unique(unlist(strsplit(c(mrna[[i]][[j]], cpg[[i]][[j]]), ";")))
  })
  names(x) <- names(mrna$JIVE)
  x
})
names(mrna.cpg) <- names(mrna)

mrna.mirna.cpg <- lapply(names(mrna), function(i){
  x <- lapply(names(mrna$JIVE), function(j){
      unique(unlist(strsplit(c(mrna[[i]][[j]], mirna[[i]][[j]], cpg[[i]][[j]]), ";")))
  })
  names(x) <- names(mrna$JIVE)
  x
})
names(mrna.mirna.cpg) <- names(mrna)
```

# Component plots

```{r multiOmicPanels_allcomponentplots, fig.path='Figures/', dev='png', fig.height = 9, fig.width = 13}
allscores <- rbind(do.call(rbind, snf_jive$pcs), do.call(rbind, snf_mofa$pcs), do.call(rbind, snf_sgcca$pcs), 
      do.call(rbind, snf_concat_splsda$pcs), do.call(rbind, snf_ensemble_splsda$pcs), 
  do.call(rbind, snf_diabloNull$pcs), do.call(rbind, snf_diabloFull$pcs)) %>% 
  as.data.frame %>% 
  mutate(Disease = rep(rep(names(snf_group), sapply(snf_group, length)), 7))

allCompPlots <- filter(allscores, Disease != "colon") %>% 
  ggplot(aes(x = PC1, y = PC2, group = pheno, color = pheno)) + geom_point() +
    facet_wrap(Disease ~ Method, scales = "free", ncol = 7) +
    stat_ellipse(level = 0.80) +
  customTheme(sizeStripFont = 10, xAngle = 0, hjust = 0.5, 
              vjust = 0.5, xSize = 10, ySize = 10, xAxisSize = 10, yAxisSize = 10) +
  xlab("Component 1") + ylab("Component 2")
allCompPlots
```

## colon component plot

```{r multiOmicsPanels_compPlot_colon, fig.path='Figures/', dev='png', fig.height = 3.5, fig.width = 14}
compPlot <- allscores %>% 
  filter(Disease == "colon") %>% 
  mutate(Method = factor(Method, c("DIABLO_full", "DIABLO_null", "Ensemble", "Concatenation", "sGCCA", "MOFA", "JIVE"))) %>% 
  ggplot(aes(x = PC1, y = PC2, group = pheno, color = pheno)) + geom_point() +
    facet_wrap(Disease ~ Method, scales = "free", ncol = 7) +
    stat_ellipse(level = 0.80) +
  customTheme(sizeStripFont = 10, xAngle = 0, hjust = 0.5, 
              vjust = 0.5, xSize = 10, ySize = 10, xAxisSize = 10, yAxisSize = 10) +
  theme(rect = element_rect(fill = "transparent")) +
  xlab("Component 1") +
  ylab("Component 2")
compPlot
```

# Silhouette coeffients per datasets per phenotype

```{r}
# function that calculates the silhouette coefficient based on a known cluster (i.e. protocol - batch or cell line)
# calculates silhouette width average
calc.sil = function(
 x, # the PC variates
 y1, y2 = NULL, # factor of interest, e.g. known batch info or known cell type
 name.y1, name.y2 = NULL # character of the factor of interest
){
 library(cluster)
 # calculate the distance, here euclidean is appropriate for PCA, NOT for t-SNE
 dist.res = daisy(x, metric = 'euclidean')
 # for factor 1
 sil.batch.res1 = silhouette(x = as.numeric(y1), dist = dist.res)
 # if factor 2 is provided
 if(!is.null(y2))  sil.batch.res2 = silhouette(x = as.numeric(y2), dist = dist.res)
 
 # extract average width silhouette per level
 res1 = c(summary(sil.batch.res1)["clus.avg.widths"]$clus.avg.widths)
 names(res1) = levels(y1)
 if(!is.null(y2)){
   res2 = c(summary(sil.batch.res2)["clus.avg.widths"]$clus.avg.widths)
   names(res2) = levels(y2)
 }
 
 # output data for plotting
 if(!is.null(y2)){
   silh.coeff = c(res1, res2)
   cluster = c(levels(y1), levels (y2))
   type = c(rep(name.y1, nlevels(y1)), rep(name.y2, nlevels(y2)))
 }else{
   silh.coeff = c(res1)
   cluster = c(levels(y1))
   type = rep(name.y1, nlevels(y1))
 }

 data.plot = data.frame(silh.coeff, cluster, type)

 
 return(invisible(data.plot))
}

```

```{r}
allscores <- rbind(do.call(rbind, snf_jive$pcs), do.call(rbind, snf_mofa$pcs), do.call(rbind, snf_sgcca$pcs), 
      do.call(rbind, snf_concat_splsda$pcs), do.call(rbind, snf_ensemble_splsda$pcs), 
  do.call(rbind, snf_diabloNull$pcs), do.call(rbind, snf_diabloFull$pcs)) %>% 
  as.data.frame %>% 
  mutate(Disease = rep(rep(names(snf_group), sapply(snf_group, length)), 7))

allscores %>% 
  group_by(Disease, Method) %>% 
  nest() %>% 
  mutate(sil = purrr::map(data, ~{
    calc.sil(x = as.matrix(.[, c("PC1", "PC2")]), y1 = factor(.$pheno), y2 = NULL, name.y1 ="pheno", name.y2 = NULL)
    })) %>% 
  unnest(sil) %>% 
  mutate(learning = ifelse(Method %in% c("JIVE", "MOFA", "sGCCA"), "Unsupervised", "Supervised")) %>% 
  ggplot(aes(x = Method, y = silh.coeff, fill=cluster, color=cluster)) +
  geom_bar(stat = "identity", position = "dodge") +
  facet_grid(learning~Disease) +
  customTheme(sizeStripFont = 15, xAngle = 30, hjust = 1, vjust = 1, xSize = 10, ySize = 10, xAxisSize = 10, yAxisSize = 10) +
  ylab("Silhouette coefficient")

```


# Overlap in panels

## Colon

### Intersection plot

```{r colon_overlap, fig.path='Figures/', dev='png', fig.height = 3.5, fig.width = 7}
colon_panels <-  list(JIVE = unlist(snf_jive$panel$colon), 
                   MOFA = unlist(snf_mofa$panel$colon), 
                   sGCCA = unlist(snf_sgcca$panel$colon),
                   Concatenation = unlist(snf_concat_splsda$panel$colon),
                   Ensemble = unlist(snf_ensemble_splsda$panel$colon),
                   DIABLO_null = unlist(snf_diabloNull$panel$colon),
                   DIABLO_full = unlist(snf_diabloFull$panel$colon))

colonInput <- fromList(colon_panels)
metadata <- data.frame(approaches=colnames(colonInput))
metadata$type <- "supervised"
metadata$type[metadata$approaches %in% c("JIVE","sGCCA","MOFA")] <- "unsupervised"

upset(colonInput, sets = colnames(colonInput), keep.order = TRUE,
    queries = list(list(query = intersects, params = list("Concatenation","DIABLO_null","Ensemble"), active = TRUE, color = "#56B4E9"),
      list(query = intersects, params = list("JIVE","MOFA","sGCCA", "DIABLO_full"), active = TRUE, color = "#D55E00")),
  set.metadata = list(data = metadata, plots = list(list(type = "matrix_rows", 
    column = "type", colors = c(supervised = superCol, unsupervised = unsuperCol), 
    alpha = 0.5))))
grid.text("Colon",x = 0.65, y=0.95, gp=gpar(fontsize=20))
```

### Venn diagram

```{r}
venn(colon_panels, zcolor = "style", cexsn = 1, cexil = 1.3)
```

## Kidney

### Intersection plot

```{r kidney_overlap, fig.path='Figures/', dev='png', fig.height = 3.5, fig.width = 7}
kidney_panels <-  list(JIVE = unlist(snf_jive$panel$kidney), 
                   MOFA = unlist(snf_mofa$panel$kidney), 
                   sGCCA = unlist(snf_sgcca$panel$kidney),
                   Concatenation = unlist(snf_concat_splsda$panel$kidney),
                   Ensemble = unlist(snf_ensemble_splsda$panel$kidney),
                   DIABLO_null = unlist(snf_diabloNull$panel$kidney),
                   DIABLO_full = unlist(snf_diabloFull$panel$kidney))

kidneyInput <- fromList(kidney_panels)
metadata <- data.frame(approaches=colnames(kidneyInput))
metadata$type <- "supervised"
metadata$type[metadata$approaches %in% c("JIVE","sGCCA","MOFA")] <- "unsupervised"

upset(kidneyInput, sets = colnames(kidneyInput),  keep.order = TRUE,
    queries = list(list(query = intersects, params = list("Concatenation","DIABLO_null","Ensemble"), active = TRUE, color = "#56B4E9"),
      list(query = intersects, params = list("JIVE","MOFA","sGCCA", "DIABLO_full"), active = TRUE, color = "#D55E00")),
  set.metadata = list(data = metadata, plots = list(list(type = "matrix_rows", 
    column = "type", colors = c(supervised = superCol, unsupervised = unsuperCol), 
    alpha = 0.5))))
grid.text("Kidney",x = 0.65, y=0.95, gp=gpar(fontsize=20))

```

### Venn diagram

```{r}
venn(kidney_panels, zcolor = "style")
```

## GBM

### Intersection plot

```{r gbm_overlap, fig.path='Figures/', dev='png', fig.height = 3.5, fig.width = 7}
gbm_panels <-  list(JIVE = unlist(snf_jive$panel$gbm), 
                   MOFA = unlist(snf_mofa$panel$gbm), 
                   sGCCA = unlist(snf_sgcca$panel$gbm),
                   Concatenation = unlist(snf_concat_splsda$panel$gbm),
                   Ensemble = unlist(snf_ensemble_splsda$panel$gbm),
                   DIABLO_null = unlist(snf_diabloNull$panel$gbm),
                   DIABLO_full = unlist(snf_diabloFull$panel$gbm))

gbmInput <- fromList(gbm_panels)
metadata <- data.frame(approaches=colnames(gbmInput))
metadata$type <- "supervised"
metadata$type[metadata$approaches %in% c("JIVE","sGCCA","MOFA")] <- "unsupervised"

upset(gbmInput, sets = colnames(gbmInput), keep.order = TRUE,
    queries = list(list(query = intersects, params = list("Concatenation","DIABLO_null","Ensemble"), active = TRUE, color = "#56B4E9"),
      list(query = intersects, params = list("JIVE","MOFA","sGCCA", "DIABLO_full"), active = TRUE, color = "#D55E00")),
  set.metadata = list(data = metadata, plots = list(list(type = "matrix_rows", 
    column = "type", colors = c(supervised = superCol, unsupervised = unsuperCol), 
    alpha = 0.5))))
grid.text("GBM",x = 0.65, y=0.95, gp=gpar(fontsize=20))

```

### Venn diagram

```{r}
venn(gbm_panels, zcolor = "style")
```

## Lung

### Intersection plot

```{r lung_overlap, fig.path='Figures/', dev='png', fig.height = 3.5, fig.width = 7}
lung_panels <-  list(JIVE = unlist(snf_jive$panel$lung), 
                   MOFA = unlist(snf_mofa$panel$lung), 
                   sGCCA = unlist(snf_sgcca$panel$lung),
                   Concatenation = unlist(snf_concat_splsda$panel$lung),
                   Ensemble = unlist(snf_ensemble_splsda$panel$lung),
                   DIABLO_null = unlist(snf_diabloNull$panel$lung),
                   DIABLO_full = unlist(snf_diabloFull$panel$lung))

lungInput <- fromList(lung_panels)
metadata <- data.frame(approaches=colnames(lungInput))
metadata$type <- "supervised"
metadata$type[metadata$approaches %in% c("JIVE","sGCCA","MOFA")] <- "unsupervised"

upset(lungInput, sets = colnames(lungInput), keep.order = TRUE,
    queries = list(list(query = intersects, params = list("Concatenation","DIABLO_null","Ensemble"), active = TRUE, color = "#56B4E9"),
      list(query = intersects, params = list("JIVE","MOFA","sGCCA", "DIABLO_full"), active = TRUE, color = "#D55E00")),
  set.metadata = list(data = metadata, plots = list(list(type = "matrix_rows", 
    column = "type", colors = c(supervised = superCol, unsupervised = unsuperCol), 
    alpha = 0.5))))
grid.text("Lung",x = 0.65, y=0.95, gp=gpar(fontsize=20))

```

### Venn diagram

```{r}
venn(lung_panels, zcolor = "style")
```

# Gene set enrichment analysis

Finally, we carried out gene set enrichment analysis on each multi-omics biomarker panel (using gene symbols of mRNAs and CpGs) against 10 gene set collections (see Methods) and tabulated the number of significant (FDR=5%) gene sets. The DIABLO_full model identified the greatest number of significant gene sets across the 10 gene set collections and generally ranked higher than the other methods in the colon (7 collections), gbm (5 collections) and lung (5 collections) cancer datasets, whereas JIVE outperformed all other methods in the kidney cancer datasets (6 collections). Unlike all other approaches considered, DIABLO_full, which aimed to explain both the correlation structure between multiple omics layers and a phenotype of interest, implicated the greatest number of known biological gene sets (pathways/functions/processes etc.). 

  * We wished to assess the enrichment of the selected features across a variety of annotated gene sets in the MSigDB collection (http://software.broadinstitute.org/gsea/msigdb), in particular:
1. C1 - positional gene sets  for each human chromosome and cytogenetic band.
2. C2 – curated gene sets (Pathway Interaction DB [PID], Biocarta [BIOCARTA], Kyoto Encyclopedia of Genes and Genomes [KEGG], Reactome [REACTOME], and others)
3. C3 - 	motif gene sets  based on conserved cis-regulatory motifs from a comparative analysis of the human, mouse, rat, and dog genomes.
4.	C4 – computational gene sets (from the Cancer Gene Neighbourhoods [CGN] and Cancer Modules [CM] – citation available via: http://www.broadinstitute.org/gsea/msigdb/collections.jsp)
5. C5 - GO gene sets  consist of genes annotated by the same GO terms.
6.	C6 – ontologic gene sets (Gene sets represent signatures of cellular pathways which are often dis-regulated in cancer).
7. C7 - immunologic gene sets  defined directly from microarray gene expression data from immunologic studies.
8. H - hallmark gene sets  are coherently expressed signatures derived by aggregating many MSigDB gene sets to represent well-defined biological states or processes.
&
A. BTM - Blood Transcriptional Modules (https://www.ncbi.nlm.nih.gov/pmc/articles/PMC2727981/)
B. TISSUES - cell-specific expression from Benita et al. Blood 2008 http://www.bloodjournal.org/content/115/26/5376

  *  Significance of enrichment was determined using a hypergeometric test of the overlap between the selected features (mapped to official HUGO gene symbols or official miRNA symbols) and the various gene sets contained in the collections. Resulting p-values were corrected for multiple hypothesis using the Benjamini-Hochberg procedure applied across ALL genesets (10k+ tests – as pessimistic as possible). Adjusted p-values are reported in the fdr column.

## All gene sets
  * only using mRNA and CpGs

```{r gensetEnrichmentResults}
enriched_sets <- rapply(mrna.cpg, function(i){
    sear(i, "mrna") %>% 
    dplyr::group_by(collection) %>% 
    dplyr::summarise(sig = sum(fdr < 0.05)) %>% 
    dplyr::filter(collection != "ARCHIVED")
}, how = "list") %>% lapply(., function(i){
  do.call(rbind, i) %>% 
    mutate(disease = rep(names(i), each = nrow(i[[1]])))
})
enrichedPathways <- do.call(rbind, enriched_sets) %>% 
    mutate(method = rep(names(enriched_sets), each = nrow(enriched_sets[[1]])))
enrichedPathways$type <- "supervised"
enrichedPathways$type[enrichedPathways$method %in% c("JIVE", "MOFA", "sGCCA")] <- "unsupervised"

kable(enrichedPathways %>% 
  spread(collection, sig))

sigPathways_cancer_db <- enrichedPathways %>%
  dplyr::select(-type) %>% 
  spread(method, sig) %>% 
  dplyr::select(disease, collection, JIVE, MOFA, sGCCA, Concatenation, Ensemble, DIABLO_null, DIABLO_full) %>% 
  arrange(disease, collection)
sigPathways_cancer <- sigPathways_cancer_db %>% 
  group_by(disease) %>% 
  dplyr::summarise(JIVE=sum(JIVE), MOFA=sum(MOFA), sGCCA=sum(sGCCA), Concatenation=sum(Concatenation), Ensemble=sum(Ensemble), DIABLO_null=sum(DIABLO_null), DIABLO_full=sum(DIABLO_full)) %>% 
  mutate(collection = "TOTAL") %>% 
  dplyr::select(disease, collection, JIVE, MOFA, sGCCA, Concatenation, Ensemble, DIABLO_null, DIABLO_full)

rbind(sigPathways_cancer_db, sigPathways_cancer) %>% 
  arrange(disease, collection) %>% 
  write.csv(., paste0(WhereAmI, "/Tables/multiOmicPanels_biologicalEnrichment.csv"))

```

## All gene sets combined
  * only using mRNA and CpGs
  
```{r}
kable(enrichedPathways %>% 
  group_by(disease, method) %>% 
    summarise(count = sum(sig)) %>% 
    spread(disease, count))
```


### which method is leads to the greatest number of signficant pathways?

```{r}
enrichedPathways %>% 
  group_by(disease, collection) %>% 
  nest() %>% 
  mutate(method = purrr::map(data, ~{
    topMethod <- as.character(as.matrix(.[.$sig == max(.$sig), "method"]))
    ifelse(length(topMethod) == 1, topMethod, "tie")
  })) %>% 
  dplyr::select(disease, collection, method) %>% 
  unnest() %>% 
  as.data.frame() %>% 
  group_by(disease, method) %>% 
  summarise(n=n()) %>% 
  ggplot(aes(x = disease, y = n, fill = method, color = method)) +
  geom_bar(stat = "identity", position = "dodge")



enrichedPathways %>% 
  group_by(disease, collection) %>% 
  nest() %>% 
  mutate(method = purrr::map(data, ~{
    data.frame(method = .$method, 
               rank = rank(.$sig, ties.method="max")-min(rank(.$sig, ties.method="max")))
  })) %>% 
  dplyr::select(disease, collection, method) %>% 
  unnest() %>% 
  as.data.frame() %>% 
  group_by(disease, method) %>% 
  ggplot(aes(x = disease, y = rank, fill = method, color = method)) +
  geom_bar(stat = "identity", position = "dodge") +
  facet_wrap(~collection) +
  customTheme(sizeStripFont = 15, xAngle = 40, hjust = 1, vjust = 1, 
    xSize = 10, ySize = 10, xAxisSize = 10, yAxisSize = 10)



```

# Connectivity

The level of connectivity of each of the seven multi-omics panels was assessed by generating networks from the feature adjacency matrix at various Pearson correlation coefficient cut-offs. At all cut-offs, unsupervised approaches produced networks with greater connectivity (number of edges) compared to supervised approaches. In addition, biomarker panels identified by DIABLO_full, were more similar to those identified by unsupervised approaches, including high graph density, low number of communities and large number of triads, indicating that DIABLO_full identified discriminative sets of features that were tightly correlated across biological compartments. 

```{r}
cormat_pearson <- lapply(multiOmicPanels, function(i){
  mapply(function(x, y){
    do.call(cbind, y)[, unlist(x)] %>% 
      cor(., method = "pearson")
  }, x = i, y = snf_data, SIMPLIFY = FALSE)
})

```

```{r}
## datasets of multi-omic panels
data <- multiOmicPanels %>%
  purrr::map(~ {                  # for every model
    purrr::map2(., snf_data, ~ {  # for every experiment
      purrr::map2(.x, .y, ~ {     # for every list of features and list of datatypes
        .y[ , .x]
      })
    })
  })

# combine
data <- purrr::modify_depth(data, 2, purrr::reduce, cbind)
```

```{r}
## adjacency matrices
adj <- modify_depth(data, 2, cor)
```

## Number of connections

```{r multiOmicsPanels_nConnections, fig.path='Figures/', dev='png', fig.height = 5, fig.width = 12}
cor_mat <- cormat_pearson
plots <- list()

nConnectionsDat <- cor_mat %>%
  modify_depth(2, ~ tibble(cor = .[lower.tri(., diag = F)])) %>%
  purrr::map(bind_rows, .id = 'dataset') %>%
  bind_rows(.id = 'model') %>%
  mutate(cor = abs(cor)) %>% 
  group_by(model, dataset) %>% 
  mutate(cor = abs(cor)) %>% 
  summarise(cor_0.5 = sum(cor > 0.5), cor_0.65 = sum(cor > 0.55),
    cor_0.6 = sum(cor > 0.60), cor_0.65 = sum(cor > 0.65),
    cor_0.7 = sum(cor > 0.70), cor_0.75 = sum(cor > 0.75),
    cor_0.8 = sum(cor > 0.8), cor_0.85 = sum(cor > 0.85),
    cor_0.9 = sum(cor > 0.9), cor_0.95 = sum(cor > 0.95), cor_1 = sum(cor == 1)) %>% 
  gather(cor, nConnections, -c(model:dataset)) %>% 
  mutate(cor = as.numeric(sapply(strsplit(cor, "_"), function(i) i[2])))
nConnectionsDat$type <- "supervised"
nConnectionsDat$type[nConnectionsDat$model %in% c("JIVE", "MOFA", "sGCCA")] <- "unsupervised"

nconnections <- nConnectionsDat %>% 
  filter(dataset != "colon") %>% 
  mutate(label = if_else(cor == min(cor), as.character(model), NA_character_)) %>% 
  ggplot(aes(x=cor,y=nConnections,fill=type,color=type,linetype=model)) +
  geom_line(size=1)+
  facet_wrap(~dataset, nr=1) +
  customTheme(sizeStripFont = 15, xAngle = 0, hjust = 0.5, vjust = 0.5, 
              xSize = 10, ySize = 10, xAxisSize = 10, yAxisSize = 10) +
  xlab("Absolute correlation coefficient cut-off") +
  ylab("Number of edges") +
  scale_fill_manual(values=c(superCol, unsuperCol)) +
  scale_color_manual(values=c(superCol, unsuperCol)) +
  geom_label_repel(aes(label = label),color="black",
                  nudge_x = 0,
                  na.rm = TRUE)
nconnections

```

### Colon only

```{r colon_connectivity, fig.path='Figures/', dev='png', fig.height = 5, fig.width = 8}
nConnectionsDat %>% 
  filter(dataset == "colon") %>% 
  mutate(label = if_else(cor == min(cor), as.character(model), NA_character_)) %>% 
  ggplot(aes(x=cor,y=nConnections,fill=type,color=type,linetype=model)) +
  geom_line(size=1)+
  facet_wrap(~dataset, nr=1) +
  customTheme(sizeStripFont = 15, xAngle = 0, hjust = 0.5, vjust = 0.5, 
              xSize = 10, ySize = 10, xAxisSize = 10, yAxisSize = 10) +
  xlab("Absolute correlation coefficient cut-off") +
  ylab("Number of edges") +
  scale_fill_manual(values=c(superCol, unsuperCol)) +
  scale_color_manual(values=c(superCol, unsuperCol)) +
  geom_label_repel(aes(label = label),color="black",
                  nudge_x = 0,
                  na.rm = TRUE) +
  theme(legend.position = c(0.75, 0.7))
```

# Network attributes

```{r multiOmicsPanels_networkAttributes, fig.path='Figures/', dev='png', fig.height = 10, fig.width = 10}
df_list <- list()
corSeq <- seq(0.5, 0.95, 0.05)
for(i in 1:length(corSeq)){
  cor_cutoff <- corSeq[i]
  df <- cor_mat %>%
  modify_depth(2, ~ tibble(adj = I(list(.)))) %>%
  purrr::map(bind_rows, .id = 'data') %>%
  bind_rows(.id = 'model')

# graphs
df$net <- df$adj%>%
  purrr::map(~ {
    .[abs(.) < cor_cutoff] <- 0
    .[abs(.) >= cor_cutoff] <- 1
    .
  }) %>%
  purrr::map(igraph::graph_from_adjacency_matrix, mode = 'lower', weighted = NULL, diag = F)

## attributes
df$clusters <- purrr::map(df$net, igraph::cluster_edge_betweenness)
df$ncommunity <- purrr::map(df$clusters, length) %>% unlist()
#df$modularity <- purrr::map(df$clusters, modularity) %>% unlist()
#df$transitivity <- purrr::map(df$net, igraph::transitivity) %>% unlist()
df$triads <- purrr::map(df$net, igraph::triad_census) %>% purrr::map(last) %>% unlist()
df$density <- purrr::map(df$net, igraph::edge_density) %>% purrr::map(last) %>% unlist()
df$cor_cutoff <- cor_cutoff
df_list[[i]] <- df

}

attributesDat <- do.call(rbind, df_list) %>% 
  gather(attributes, value, ncommunity:density)
attributesDat$type <- "supervised"
attributesDat$type[attributesDat$model %in% c("JIVE", "MOFA", "sGCCA")] <- "unsupervised"

netAttributes <- attributesDat %>% 
   mutate(label = if_else(cor_cutoff == min(cor_cutoff), as.character(model), NA_character_)) %>% 
  ggplot(aes(x = cor_cutoff, y = value, fill = type, color = type, linetype = model)) +
  geom_point()+
  geom_line()+
  facet_grid(attributes ~ data, scales = "free") +
  customTheme(sizeStripFont = 15, xAngle = 0, hjust = 0.5, vjust = 0.5, 
               xSize = 10, ySize = 10, xAxisSize = 10, yAxisSize = 10) +
  scale_fill_manual(values=c(superCol, unsuperCol)) +
  scale_color_manual(values=c(superCol, unsuperCol)) +
  geom_label_repel(aes(label = label),color="black",
                  nudge_x = 0,
                  na.rm = TRUE)
netAttributes

```

## all network plots

```{r multiOmicsPanels_networks, fig.path='Figures/', dev='png', fig.height = 10, fig.width = 15}
df <- cor_mat %>%
modify_depth(2, ~ tibble(adj = I(list(.)))) %>%
purrr::map(bind_rows, .id = 'data') %>%
bind_rows(.id = 'model')

# graphs
df$net <- df$adj%>%
  purrr::map(~ {
    .[abs(.) < 0.25] <- 0
    .[abs(.) >= 0.25] <- 1
    .
  }) %>%
  purrr::map(igraph::graph_from_adjacency_matrix, mode = 'lower', weighted = NULL, diag = F)
df$clusters <- purrr::map(df$net, igraph::cluster_edge_betweenness)

df$ggraphs <- pmap(list(c = df$clusters, g = df$net) , function(c, g) {
  V(g)$community <- as.character(c$membership)
  V(g)$block <- sapply(strsplit(V(g)$name, "_"), function(i) i[1])
  ggraph(g, layout = 'igraph', algorithm = 'nicely') +
    # geom_edge_fan(colour = 'lightgrey', show.legend = FALSE) +
    geom_encircle(aes(x = x, y = y, group = community),
                  s_shape = 0.5, expand = 0.025, colour = 'lightgrey') +
    geom_node_point(aes(fill = block), shape = 21, colour = 'white', size = 4) +
    scale_x_continuous(expand = c(0.25, 0.25)) +
    scale_y_continuous(expand = c(0.25, 0.25)) +
    theme(aspect.ratio = 1.25,
          # legend.position = 'bottom',
          axis.ticks = element_blank(),
          axis.text = element_blank(),
          axis.line = element_blank(),
          axis.title = element_blank())
})

# grab legend
legend <- get_legend(df$ggraphs[[1]])

# r1 <- plots$adjacency
nets <- lapply(unique(df$data), function(i){
  r3 <- filter(df, data == i)
  r3 <- map2(r3$ggraphs, r3$model, ~ .x + ggtitle(.y) + theme(legend.position = 'none'))
  r3 <- rev(r3)
  r3 <- plot_grid(
  r3[[1]],
  r3[[2]],
  r3[[3]],
  r3[[4]],
  r3[[5]],
  r3[[6]],
  r3[[7]],
  legend,
  nrow = 1, rel_widths = c(1, 1, 1, 1, 1, 1, 1, 0.4)
)
r3
})
names(nets) <- unique(df$data)

p <- plot_grid(nets$gbm, nets$kidney, nets$lung, nrow = 3, ncol = 1, labels = c('', ''))
p

```

## network and component plot of the multi-omic panels derived using the colon cancer dataset

The plots below depict the networks of all multi-omics biomarker panels for the colon cancer dataset, which show higher modularity (a limited number of large clusters of variables; circled) for the DIABLO_full and the unsupervised approaches as compared to the supervised ones. The corresponding component plots show a clear separation between the high and low survival groups for the panels derived using supervised approaches, whereas the unsupervised approaches could not segregate the survival groups. 

### networks

```{r multiOmicsPanels_networks_colon, fig.path='Figures/', dev='png', fig.height = 6, fig.width = 14}
nets$colon
```

## References

3. Wang B, Mezlini AM, Demir F, Fiume M, Tu Z, Brudno M, et al. Similarity network fusion for aggregating data types on a genomic scale. Nat Methods [Internet]. 2014 [cited 2016 Jan 19];11:333–7. Available from: http://www.nature.com/doifinder/10.1038/nmeth.2810
14. Lê Cao K-A, Boitard S, Besse P. Sparse PLS discriminant analysis: biologically relevant feature selection and graphical displays for multiclass problems. BMC Bioinformatics [Internet]. 2011 [cited 2015 Jul 15];12:253. Available from: http://www.biomedcentral.com/1471-2105/12/253/
15. Tenenhaus A, Philippe C, Guillemot V, Le Cao K-A, Grill J, Frouin V. Variable selection for generalized canonical correlation analysis. Biostatistics [Internet]. 2014 [cited 2015 Jul 15];15:569–83. Available from: http://biostatistics.oxfordjournals.org/cgi/doi/10.1093/biostatistics/kxu001
19. The TCGA Research Network. The Cancer Genome Atlas [Internet]. Available from: http://cancergenome.nih.gov/
20. Singh A, Yamamoto M, Kam SHY, Ruan J, Gauvreau GM, O’Byrne PM, et al. Gene-metabolite expression in blood can discriminate allergen-induced isolated early from dual asthmatic responses. Hsu Y-H, editor. PLoS ONE [Internet]. 2013 [cited 2015 Jul 18];8:e67907. Available from: http://dx.plos.org/10.1371/journal.pone.0067907
21. Singh A, Yamamoto M, Ruan J, Choi JY, Gauvreau GM, Olek S, et al. Th17/Treg ratio derived using DNA methylation analysis is associated with the late phase asthmatic response. Allergy Asthma Clin Immunol [Internet]. 2014 [cited 2016 Mar 2];10:32. Available from: http://www.biomedcentral.com/content/pdf/1710-1492-10-32.pdf
23. Liberzon A, Birger C, Thorvaldsdóttir H, Ghandi M, Mesirov JP, Tamayo P. The Molecular Signatures Database Hallmark Gene Set Collection. Cell Syst [Internet]. 2015 [cited 2018 Jan 30];1:417–25. Available from: http://linkinghub.elsevier.com/retrieve/pii/S2405471215002185
